lib/auth/vnetconfig/v1/service.go

```diff
@@ -0,0 +1,201 @@
+type Service struct {
+	// Opting out of forward compatibility, this service must implement all service methods.
+	vnet.UnsafeVnetConfigServiceServer
+
+	storage    *local.VnetConfigService
+	authorizer authz.Authorizer
+}
+
+func NewService(storage *local.VnetConfigService, authorizer authz.Authorizer) *Service {
+	return &Service{
+		storage:    storage,
+		authorizer: authorizer,
+	}
+}
+
+func (s *Service) GetVnetConfig(ctx context.Context, _ *vnet.GetVnetConfigRequest) (*vnet.VnetConfig, error) {
+	authCtx, err := s.authorizer.Authorize(ctx)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if err := authCtx.CheckAccessToKind(types.KindVnetConfig, types.VerbRead); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	vnetConfig, err := s.storage.GetVnetConfig(ctx)
+	if err != nil {
+		return nil, utils.OpaqueAccessDenied(err)
+	}
+
+	if err := checkAccessToResource(authCtx, vnetConfig, types.VerbRead); err != nil {
+		return nil, utils.OpaqueAccessDenied(err)
+	}
+
+	return vnetConfig, nil
+}
+
+func (s *Service) CreateVnetConfig(ctx context.Context, req *vnet.CreateVnetConfigRequest) (*vnet.VnetConfig, error) {
+	authCtx, err := s.authorizer.Authorize(ctx)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if err := checkAccessToResource(authCtx, req.VnetConfig, types.VerbCreate); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if err := authCtx.AuthorizeAdminAction(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	vnetConfig, err := s.storage.CreateVnetConfig(ctx, req.VnetConfig)
+	return vnetConfig, trace.Wrap(err)
+}
+
+func (s *Service) UpdateVnetConfig(ctx context.Context, req *vnet.UpdateVnetConfigRequest) (*vnet.VnetConfig, error) {
+	authCtx, err := s.authorizer.Authorize(ctx)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if err := checkAccessToResource(authCtx, req.VnetConfig, types.VerbUpdate); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if err := authCtx.AuthorizeAdminAction(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	oldVnetConfig, err := s.storage.GetVnetConfig(ctx)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if err := checkAccessToResource(authCtx, oldVnetConfig, types.VerbUpdate); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	vnetConfig, err := s.storage.UpdateVnetConfig(ctx, req.VnetConfig)
+	return vnetConfig, trace.Wrap(err)
+}
+
+func (s *Service) UpsertVnetConfig(ctx context.Context, req *vnet.UpsertVnetConfigRequest) (*vnet.VnetConfig, error) {
+	authCtx, err := s.authorizer.Authorize(ctx)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// To upsert you must be allowed to Create and Update the new resource.
+	if err := checkAccessToResource(authCtx, req.VnetConfig, types.VerbCreate, types.VerbUpdate); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if err := authCtx.AuthorizeAdminAction(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// Try a few times to Create or Update the resource, in case we race with another Create or Update.
```

Comment:
Why don't we just really upsert the resource? 
If you want to ensure we don't race, you can use a RunWhileLocked to protect multiple routines to race against each other. 

it's probably not a bad call since this is a cluster config

Reply:
Why: I'm pretty sure this is the only way to actually get the RBAC semantics right without locks or AtomicWrite. With AtomicWrite I'd be doing a loop anyway (ConditionalUpdate is kind of AtomicWrite lite i guess)

I am being pretty pedantic about RBAC here and checking it on the existing resource, I doubt anyone will use a `where` clause on this anyway. If you don't like the loop I could just check access to the new resource and call it a day

Reply:
It might be worth it to check if the user has blanket `create` and `update` permissions on the whole resource type to skip all of this (if that's possible).

Reply:
I think fewer special cases is generally better, what do you think would make that worth it? Only thing I see is it could save a couple cpu cycles when someone updates this a couple times a month (if that)?

Reply:
you know what, supporting `where` on this singleton config resource with 2 fields is stupid, I'm going to rip all this out and just do basic verb checks